sqtom/fitting_1d.py

```diff
@@ -25,23 +25,90 @@
 """
 
 
+import warnings
 import numpy as np
+from scipy.optimize import root_scalar
 from lmfit import Minimizer, Parameters
 from sqtom.forward_solver import degenerate_pmf
 
 
+def two_schmidt_mode_guess(pd_data, sq_label="sq_", noise_fraction=0.001, atol=1e-8):
+    """Given a single mode histogram, this function generates a "physically" motivated guess for the loss, Schmidt
+    occupations and dark counts parameters.
+
+    This model is sensible only if there is a probability for an n == 2 event. If there are only probabilities for
+    n == 0 and n == 1 then the g2 will erroneously be calculated as 0 and as a result the remainder of the estimtes will
+    be nonsense.
+
+    Args:
+        pd_data (array): rectangular array with the probability mass functions of the photon events
+        sq_label (string): label for the squeezing parameters.
+
+    Returns:
+        dict: dictionary containing a set of "reasonable" model parameters. In the case that the pd_data only has
+        entries for n == 0, or n == 0 and n == 1, it returns None along with a warning message.
+    """
+    res = marginal_calcs_1d(pd_data)
+    nmean = res["n"]
+    g2 = res["g2"]
+    P0 = pd_data[0]
+    if np.allclose(nmean, 0, atol=atol):
+        warnings.warn("The average photon number is zero")
+        return None
+    if np.allclose(g2, 0, atol=atol):
+        warnings.warn("Not enough samples for suitable guess")
+        return None
+
+    def findeta(eta, nmean, g2, P0):
+        a = nmean / 4
+        b = (3 - g2) * (nmean ** 2) / 4 - nmean
+        c = (g2 - 3) * nmean ** 2
+        d = (3 - g2) * nmean ** 2 + 2 * nmean + 1 - 1 / P0 ** 2
+        return a * eta ** 3 + b * eta ** 2 + c * eta + d
+
+    eta_set = np.linspace(-0.01, 1.01, num=509)
+    function_root_search = np.array([findeta(i, nmean, g2, P0) for i in eta_set])
+    indices = np.array([])
+    for i in range(function_root_search.size - 1):
+        if function_root_search[i + 1] / function_root_search[i] < 0:
+            indices = np.append(indices, i)
+            indices = np.append(indices, i + 1)
+    eta = root_scalar(
+        findeta,
+        args=(nmean, g2, P0),
+        bracket=(eta_set[int(indices[-1] - 1)], eta_set[int(indices[-1])]),
+    ).root
+    eta = min(eta, 1)
+    eta = max(eta, 0)
+    d = (g2 - 2) * nmean ** 2 - eta * nmean
+    if d >= 0 and eta > 0:
+        n0 = (nmean + np.sqrt(d)) / (2 * eta)
+        n1 = (nmean - np.sqrt(d)) / (2 * eta)
+    elif d < 0 < eta:
+        n0 = nmean / (2 * eta)
+        n1 = nmean / (2 * eta)
+    else:
+        n0 = 0
+        n1 = 0
+    noise = nmean * noise_fraction
+    return {
+        "eta": eta,
+        sq_label + "0": n0,
+        sq_label + "1": n1,
+        "noise": noise,
+        "n_modes": 2,
```

Comment:
Just curious: would it be essential to include this as part of the dictionary? Wondering as the function name itself seems to incorporate it. It can be nice to have if a more general version is intended that is parameterized using the number of modes.

Reply:
Hey Antal --- This is because this function is meant to integrate nicely with `fit_1d`, for which the `guess` is a dictionary that need to have a key with the number of  modes `n_modes`.

Reply:
Sounds good, thanks!